src/lib/SaferNames/Syntax.hs

```diff
@@ -154,135 +157,62 @@ type Con n = PrimCon (Atom n)
 type Op  n = PrimOp  (Atom n)
 type Hof n = PrimHof (Atom n)
 
+type IndexStructure = Nest Binder
+
+type AtomSubstVal = SubstVal TypedBinderInfo Atom :: E -> E
+
 -- === top-level modules ===
 
 type SourceName = T.Text
 
 data SourceNameMap n = SourceNameMap
-  { fromSourceNameMap :: M.Map SourceName (Name n)}
+  { fromSourceNameMap :: M.Map SourceName (Name UnitE n)}
 
 data Module n where
   Module :: IRVariant
          -> Nest Decl n l
-         -> BindingsFrag l l'
+         -> ScopeFrag l l'
          -> SourceNameMap l'
          -> Module n
 
-type WithBindings = Abs BindingsFrag :: E -> E
-type EvaluatedModule = WithBindings SourceNameMap  :: E
-
-emptyEvaluatedModule :: EvaluatedModule n
-emptyEvaluatedModule = Abs (RecEnvFrag emptyEnv) mempty
-
-type IndexStructure = Nest Binder
-
-getBaseMonoidType :: Scope n -> Type n -> Type n
-getBaseMonoidType scope ty = case ty of
-  TabTy i b -> case projectNamesL scope (i@>()) b of
-    Just b' -> b'
-    Nothing -> error "Can't use a dependent table as a monoid"
-  _         -> ty
+data EvaluatedModule (n::S) where
+  EvaluatedModule :: ScopeFrag n l -> SourceNameMap l -> EvaluatedModule n
 
 -- === effects ===
 
-data EffectRow n = EffectRow (S.Set (Effect n)) (Maybe (Name n))
-                   deriving Show
+data EffectRow n = EffectRow (S.Set (Effect n)) (Maybe (AtomName n))
```

Comment:
Should the extensions of effect rows really be atom names?

Reply:
Good question. I tried playing with a different flavor of name but `AtomName` ended up making the most sense. Unlike special names like `Name DataDef`, the name of an effect row extension can be bound by an ordinary lambda or pi binder.